keps/sig-auth/20190607-certificates-api.md

```diff
@@ -235,6 +236,10 @@ Cluster admins can either:
 2. grant signer-specific approval permissions using the bootstrap roles starting in 1.18
 3. disable the approval-authorizing admission plugin in 1.18 (if they don't care about partitioning approver rights)
  
+#### Supporting intermediate certificates.
+It's important that the Certificates API fully supports using an intermediate certificate as the signer, including the case where the signer is itself signed by an intermediate. As a result, the user must be able to specify that the signing certificates are provided to the client, in which case their PEM values would be appended to the Certificate in the CertificateSigningRequestStatus. Conforming clients will parse the full chain of certificates provided by the apiserver and present them all for new TLS handshakes.
+
+Firstly, we will add a new bool flag to the controller, `cluster-signing-include-signers`, which requires that the signing certificate is always appended in this way. Currently the controller only allows a single signing cert in `cluster-signing-cert-file`, so this approach would only work in the case where the signing certificate is a child of the root CA, in which case it is the only certificate that needs to be provided to the client. To support further nested signing certificates, we additionally need to relax the contraint of the cert file parameter to accept multiple signers in a chain. The additional certificates would only be used if the include signers option is set; they are not needed in signing.
```

Comment:
nit: if you wrap at 80/120 LoC, it makes it a little easier to review/discuss :)

> The additional certificates would only be used if the include signers option is set; they are not needed in signing.

It seems non-obvious to have the `--cluster-signing-cert-file` parameter specifically say that "some of these things actually won't be used for signing".

I may have missed something in previous discussion, but what's the rationale behind a bool vs an explicit `--cluster-signing-cert-chain` flag which points to a file containing the chain expected to be appended to the certificates the signer signs? I know this was talked about before and I thought the consensus was on an explicit file/flag.

To try and get an idea of what either configuration would look like either way...

With the bool:

* single self signed root: configure as we do today (no `--cluster-signing-include-signers` flag
* root -> intermediate: `--cluster-signing-cert-file=<path-to-intermediate>` `--cluster-signing-include-signers=true`
* root -> intermediate -> intermediate: `--cluster-signing-cert-file=<path-to-file-with-both-intermediates>` `--cluster-signing-include-signers=true`

& with the additional file flag:

* single self signed root: configure as we do today (no `--cluster-signing-chain-file` flag
* root -> intermediate: `--cluster-signing-cert-file=<path-to-intermediate>` `--cluster-signing-chain-file=<path-to-intermediate>`
* root -> intermediate1 -> intermediate2: `--cluster-signing-cert-file=<path-to-intermediate2-file>` `--cluster-signing-chain-file=<path-to-file-with-both-intermediates>`

---

I think the `bool` option isn't actually that unclear for anyone with some PKI background (who is familiar with concatenating PEM files together), although the fact that `--cluster-signing-key-file` and `--cluster-signing-cert-file` now are a bit ambiguous could definitely cause confusion (one of these can contain multiple certificates, the other contains a private key that *must* be valid with the *first* intermediate only).

The additional `file` option seems a bit more verbose, but at least it is explicit (and validation in `kube-controller-manager` could check to ensure that only a single PEM is included in the `--cluster-signing-cert-file` too)

More of a brain dump than a concrete answer/review :)

Reply:
I like the file option too, but the spanner in the works for me was:
1. In probably(?) the most common case where you're not signing with a root, you have root -> intermediate. In that case, you provide the same file to two different parameters, which just feels weird to me as an end user
2. As an addendum to that, managing rotation of these files becomes more complicated; we now have two files that can contain overlapping content or may be the same file, where we're generally interested in the deduplicated sum of both files. We can't set up individual file rotation controllers on them, we'd need to do something bespoke to watch them together, although there will technically always be a race condition that needs to be handled

It just feels simpler to me to have all the certs in one place, and let people decide whether to include them in the cert response. To be honest, I'd advocate to include signing certs by default, maybe filtering out any root certs, but its not as explicit.